controllers/agents/svc/serviceclass_controller.go

```diff
@@ -93,7 +102,45 @@ func (r *ServiceClassReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	// then, write all the registered services up to the primaza cluster
-	err = r.CreateRegisteredServices(ctx, &serviceClass, *services)
+	var handler HandleFunc
+	if serviceClass.DeletionTimestamp.IsZero() {
+		handler = func(remote_client client.Client, rs v1alpha1.RegisteredService) error {
+			op, err := controllerutil.CreateOrUpdate(ctx, remote_client, &rs, func() error { return nil })
+			if err != nil {
+				reconcileLog.Error(err, "Failed to create registered service",
+					"service", rs.Name,
+					"namespace", rs.Namespace)
+			} else {
+				reconcileLog.Info("Wrote registered service", "service", rs.Name, "namespace", rs.Namespace, "operation", op)
+			}
+			return err
+		}
+
+		// add a finalizer since we have deletion logic
+		if controllerutil.AddFinalizer(&serviceClass, finalizer) {
+			if err = r.Update(ctx, &serviceClass, &client.UpdateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else if controllerutil.ContainsFinalizer(&serviceClass, finalizer) {
+		handler = func(remote_client client.Client, rs v1alpha1.RegisteredService) error {
+			if err := remote_client.Delete(ctx, &rs); err != nil {
+				reconcileLog.Error(err, "Failed to delete registered service", "namespace", rs.Namespace)
+			}
+			return err
+		}
+
+		// remove the finalizer so we don't requeue
+		if controllerutil.RemoveFinalizer(&serviceClass, finalizer) {
+			if err = r.Update(ctx, &serviceClass, &client.UpdateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
as the handler is executed at L143, the finalizer will be removed even if handler fails. Also deletion handler func should check if it can ignore the error (i.e. NotFoundError) or not. If not, e.g. network error, we should retry before deleting the finalizer.

Would it be a solution to move this logic into the handler?

Reply:
I've changed this logic, PTAL.

Reply:
Logic lgtm, but code needs formatting

Reply:
should be fixed now